pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -353,10 +571,13 @@ impl<'a> VectoredBlobReader<'a> {
                 )
             };
 
-            let start_raw = offset_in_buf + size_length;
-            let end_raw = match next {
-                Some((next_blob_start_offset, _)) => next_blob_start_offset - start_offset,
-                None => start_raw + blob_size,
+            let start_raw = blob_start_in_buf + size_length;
+            let end_raw = {
+                if *blob_start == last_blob_start {
+                    start_raw + blob_size
```

Comment:
Why is this necessary? Is the `blob_end` not valid in the last element of `blobs_at`?

It seems to me this is a sort of pattern-matched port of Some / None logic that we had with the `.zip()` before this PR.

Reply:
> Why is this necessary? Is the blob_end not valid in the last element of blobs_at?

`blob_end` is indeed not valid for the last element of `blobs_at`. This is used to handle the following case, where the end offset is set to `data_offset`.

https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/storage_layer/delta_layer.rs#L968-L971

We obtain this `data_offset` from https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/storage_layer/delta_layer.rs#L1425-L1438

Note L1427: 
> The file structure is summary, values, and index. We often need this for the size of last blob.

I would this is not always the case if the last blob ends earlier than the page boundary.

If we turn off the special casing, then tests will fail:

```console
--- STDERR:              pageserver tenant::storage_layer::delta_layer::test::test_delta_layer_vectored_read_end_to_end ---
thread 'tenant::storage_layer::delta_layer::test::test_delta_layer_vectored_read_end_to_end' panicked at pageserver/src/tenant/vectored_blob_io.rs:576:13:
assertion `left == right` failed
  left: 5165
 right: 100
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```




Options I see:
1. Always use parsed blob_size + remove assertions on L582. 
2. keep doing the special casing.

https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/vectored_blob_io.rs#L576-L582

Reply:
cc @VladLazar , so, the summary block is actually padded up to the next page boundary

Reply:
refactored in 70dc350ffd107acc07e4b19e7d1a292236434e0d. We will directly use start offset and parsed blob_size to determine the end offset. As a result, 

1. stopped storing `end_offset` for each blob.
2. no special casing for last blob in `read_blobs`
3. removed `assert_eq!(end_raw - start_raw, blob_size)` assertion.

cc. @VladLazar, I think we can trust the parsed blob size?